Add the ResNet_vd backbone

gowthamkpr opened the pull request:
This adds the ResNet_vd backbone to prepare for differential binarization model.

Architecture is adopted from PaddleOCR (https://github.com/PaddlePaddle/PaddleOCR/blob/main/ppocr/modeling/backbones/det_resnet_vd.py).

This is the first step of #1739:
* Aliases have been removed from ResNet_vd
* Model has been moved out of the subdirectory
* Tests (from keras_cv) have been added

divyashreepathihalli commented on keras_nlp/src/models/resnet_vd/resnet_vd_backbone_tests.py:
```diff
@@ -0,0 +1,122 @@
+        model(self.input_batch)
+
+    @pytest.mark.large  # Saving is slow, so mark these large.
+    def test_saved_model(self):
```
Use KerasNLPs standard test routine like this for the saved model test - https://github.com/keras-team/keras-nlp/blob/keras-hub/keras_nlp/src/models/vgg/vgg_backbone_test.py#L44

divyashreepathihalli commented on keras_nlp/src/models/resnet_vd/resnet_vd_backbone_tests.py:
```diff
@@ -0,0 +1,122 @@
+
+
+class ResNetVdBackboneTest(TestCase):
+    def setUp(self):
```
Also, please add a backbone basics tests like - https://github.com/keras-team/keras-nlp/blob/keras-hub/keras_nlp/src/models/vgg/vgg_backbone_test.py#L34

divyashreepathihalli reviewed (changes requested):
Thank you for the PR Gowtham! Looks good overall. 
With KerasHub if the model has multiple versions with small changes in architecture, they will be added as one model and the configuration will be handled in the args. 
We have this PR for resnet - https://github.com/keras-team/keras-nlp/pull/1765
The architecture seems similar(please let me know if this is not the case and this needs to be a separate model), can we combine ResnetVD into that one?

divyashreepathihalli commented:
Also to format the code you can use `shell/api_gem.sh`, `shell/format.sh` and `shell/lint.sh` at kerasnlp root dir. You can also run the tests locally with `pytest keras_nlp/src/models/resnet_vd`

mattdangerw commented on keras_nlp/src/models/resnet_vd/resnet_vd_backbone.py:
```diff
@@ -0,0 +1,381 @@
+        )
+        return config
+
+    @classproperty
+    def presets(cls):
+        """Dictionary of preset names and configurations."""
+        return copy.deepcopy(backbone_presets)
+
+    @classproperty
+    def presets_with_weights(cls):
+        """Dictionary of preset names and configurations that include
+        weights."""
+        return copy.deepcopy(backbone_presets_with_weights)
```
we shouldn't need to do this anymore

gowthamkpr replied:
@mattdangerw PTAL https://github.com/keras-team/keras-nlp/pull/1766#discussion_r1717086972

mattdangerw commented on keras_nlp/src/models/resnet_vd/resnet_vd_backbone_presets.py:
```diff
@@ -0,0 +1,69 @@
+            "official_name": "ResNetVd",
+            "path": "resnet_vd",
+        },
+        # "kaggle_handle": "kaggle://keras/resnetv1/keras/resnet18/2",
```
I'm confused at this. Is resnet vd just the same structure as resnet?

gowthamkpr commented:
I've changed the code to reuse `ResNetBackbone` (#1765) and to set the variant based on a parameter. For this, I replaced `use_pre_activation` with `version`, which can be either `"v1"`, `"v2"` or `"vd"`. Within the code, I retained `use_pre_activation` to retain more concise branching. This can be changed, of course.

divyashreepathihalli commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -86,7 +86,7 @@ class ResNetBackbone(Backbone):
         block_type="basic_block",
-        use_pre_activation=True,
+        version="v2",
         pooling="avg",
```
lets keep the `use_pre_activation` arg. instead of `version` lets add arg something like `use_vd_pooling` or something like that

gowthamkpr replied:
Sure, I can change this.
Just a note - with two booleans, we convey that you can set `use_pre_activation = use_vd_pooling = True`. Since both architecture variants introduce subtle modifications, it will not be entirely clear from the interface what the outcome would be. My personal preference would be having it inherently mutually exclusive.

divyashreepathihalli replied:
yes, we should have two different args!

divyashreepathihalli commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -139,21 +144,64 @@ def __init__(
                 axis=bn_axis, epsilon=1.001e-5, dtype=dtype, name="conv1_bn"
             )(x)
             x = layers.Activation("relu", dtype=dtype, name="conv1_relu")(x)
+            x = layers.Conv2D(
```
A colab verifying the numerics for this would be helpful the verify architecture correctness

gowthamkpr replied:
Modifications compared to plain ResNetV1 are
* Average pooling is used in shortcut connections instead of CNN strides
* The model's first convolutional layer is replaced with three subsequent ones
* Stride configuration in residual blocks is slightly different

[This paper](https://www.researchgate.net/publication/349646156_Automatic_Detection_and_Classification_of_Steel_Surface_Defect_Using_Deep_Convolutional_Neural_Networks) has some good visualizations of the ResNet_vd architecture.

gowthamkpr replied:
I can look into creating a notebook to verify numerics. Since we are primarily adding this for the differential binarization model, I would check numerics based on the pretrained differential binarization model from PaddleOCR unless you think we should compare numerics for the classifier network as well.

divyashreepathihalli replied:
you can verify the numerics just for the resnetvd backbone with just random weights.

divyashreepathihalli commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -209,7 +258,7 @@ def __init__(
         self.stackwise_num_strides = stackwise_num_strides
         self.block_type = block_type
-        self.use_pre_activation = use_pre_activation
+        self.version = version
```
same comment as above for changing arg name for `version`

divyashreepathihalli commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -220,7 +269,7 @@ def get_config(self):
             "stackwise_num_strides": self.stackwise_num_strides,
             "block_type": self.block_type,
-            "use_pre_activation": self.use_pre_activation,
+            "version": self.version,
```
same comment as above for changing arg name for `version`

divyashreepathihalli reviewed (changes requested):
Thanks Gowtham! I am a little confused about the added conv2d layer. A numerics verifying colab will be helpful. 
Lets update the arg as well.

james77777778 commented:
Hi @divyashreepathihalli @gowthamkpr 
Can we wait for #1769 ?
The implementation of ResNetBackbone has been changed in that PR and the numerical differences have been verified.

gowthamkpr commented:
I've updated the PR for the newest version of the code. Numerical agreement has been verified in [this notebook](https://colab.research.google.com/gist/gowthamkpr/807b5e5741cdd2c152358ee94b0136ff/resnet_vd_porting.ipynb). Observed maximum deviations from PaddleClas range between 5e-7 and 3e-6.

For verification, I have converted all ResNet_vd models from PaddleClas to Keras models. They report accuracy on ImageNet-val as shown in this figure from [this link](https://paddleclas.readthedocs.io/en/latest/models/ResNet_and_vd_en.html)
![image](https://github.com/user-attachments/assets/bfa4ee94-7823-4b3d-98c4-a3fdb9e6855f)

The accuracy readings for the ssld models (shown in red, they use knowledge distillation) are actually really good compared to the values reported in the "ResNet strikes back" paper for timm models. This might be interesting for us.

divyashreepathihalli commented:
Excellent!! can we add all of these versions for our preset? Hongyu added a built in conversion from timm for most of the resnet models - ptal - https://github.com/keras-team/keras-nlp/pull/1769

mattdangerw commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -106,6 +114,7 @@ def __init__(
         stackwise_num_strides,
         block_type,
         use_pre_activation=False,
+        use_vd_pooling=False,
```
I don't think we want this one flag that proxies a bunch of options. This is going to be hard for maintenance, and lead to confusing configs.

I think we might want to add a new block type, `vd_block`, and two new flags here. `input_conv_filters=(64,)` and `input_conv_kernel_sizes=(7,)`. Or something like that.

Then the config for resnet vd would look like...

```python
{
    ...,
    input_conv_filters=(32, 32, 64),
    input_conv_kernel_sizes=(3, 3, 3),
    block_type="vd_block",
}
```

Then you could implement the input conv layers a a loop, and implement the new vd block, by copying the `basic_block` with some minor changes.

WDYT? @james77777778 any tweaks as well?

Not sure I'm getting the exact args right, but at a high level, avoid hard coding magic numbers that define the structure of this model behind feature flags. Instead they should be directly in the config.

mattdangerw commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -155,21 +169,21 @@ def __init__(
         # Therefore, we need to manually pad the tensor.
         x = layers.ZeroPadding2D(
-            3,
+            1 if use_vd_pooling else 3,
```
Under the proposal above, this might be able to be written as `input_conv_kernel_sizes[0] // 2` or something like that?

mattdangerw commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -580,7 +678,12 @@ def apply_stack(
     """
     if name is None:
-        version = "v1" if not use_pre_activation else "v2"
+        if use_vd_pooling:
```
Similar to above I think we should remove this. The preset name is how we can track different in vd vs v2 vs v2. This will be hard to maintain.

mattdangerw reviewed (commented):
Left a few comments on how we could clean up the config here.

gowthamkpr commented:
> Excellent!! can we add all of these variants for our preset? Hongyu added a built in conversion from timm for most of the resnet models - ptal - #1769

Yes. Model files are available from the linked notebook. Both keras-nlp and PaddleClas are subject to an Apache 2.0 licence.

mattdangerw commented on keras_nlp/src/models/resnet/resnet_backbone.py:
```diff
@@ -461,7 +527,295 @@ def apply_bottleneck_block(
             use_bias=False,
             dtype=dtype,
             name=f"{name}_0_conv",
+        )(shortcut)
```
What's up with this 50 or so line addition to the basic block? Shouldn't all the changes be the VD blocks now? Am I missing something?

gowthamkpr replied:
Just as the normal ResNet, ResNet_vds use basic blocks in their small versions, and the bigger ResNet_vds use bottleneck blocks. Therefore, we need vd variants of both, `basic_block` and `bottleneck_block`.

mattdangerw replied:
Oh I think this was just github diff being weird, my bad! It looked like you were making a massive addition to the non-vd block, but I see now that's just because the code matches across both. Let's try this out!

mattdangerw reviewed (commented):
Nice looking good! I think there might be an extraneous additions to the basic block to remove.

mattdangerw reviewed (approved):
lgtm!
